btree/btree.c

```diff
@@ -1501,6 +1502,263 @@ static void generic_move(struct nd *src, struct nd *tgt,
 	node_fix(tgt, tx);
 }
 
+
+/**
+ *  -------------------------------------------------------
+ *  Section START - Credit calculation for BTree Operations
+ *  -------------------------------------------------------
+ */
+
+/*In case of fixed key, fixed value node format: max_ksize, max_vsize will be
+fixed key, value size else estimate max sizes */
+m0_bcount_t max_ksize; 
+m0_bcount_t max_vsize;
+m0_bcount_t max_node_size;
+m0_bcount_t internal_vsize; /*for internal node, value size is fix */
+
+/**
+ * node threshold can be used in btree_move_key_value_credit() to identify
+ * what fraction of node needs to be moved in case of overflow/underflow
+ */
+enum node_threshold {
+	MAX_MOVE = 70,
+	MIN_MOVE = 50,
+	ALL_MOVE = 100
+};
+
+enum {
+	BTREE_ALLOC_SHIFT = 0,
+};
+
+static void btree_mem_alloc_credit(const struct m0_btree  *btree,
+				   m0_bcount_t             size,
+				   struct m0_be_tx_credit *accum)
+{
+	m0_be_allocator_credit(NULL, M0_BAO_ALLOC_ALIGNED, size,
+			       BTREE_ALLOC_SHIFT, accum);
+}
+
+static void btree_node_alloc_credit(const struct m0_btree  *tree,
+				    struct m0_be_tx_credit *accum)
+{
+	btree_mem_alloc_credit(tree, max_node_size, accum);
+}
+
+/**
+ * Calculates the credit needed to update @nr nodes and adds this credit to
+ * @accum.
+ */
+static void btree_node_update_credit(struct m0_be_tx_credit *accum,
+				     m0_bcount_t             nr)
+{
+	struct m0_be_tx_credit cred = {};
+
+	/* @todo: why multiplication with 2 is needed? */
+	m0_be_tx_credit_mac(&cred, max_node_size, 2);
+
+	m0_be_tx_credit_mac(accum, &cred, nr);
+}
+
+/**
+ * Calculates the credit needed to create @nr nodes and adds this credit to
+ * @accum.
+ */
+static void m0_btree_create_node_credit(const struct m0_btree  *tree,
+					m0_bcount_t             nr,
+					struct m0_be_tx_credit *accum)
+{
+	struct m0_be_tx_credit cred = {};
+
+	btree_node_alloc_credit(tree, &cred);
+	/* @todo: is btree_node_update_credit needed? */
+	btree_node_update_credit(&cred, 1);
+	m0_be_tx_credit_mac(accum, &cred, nr);
+}
+
+/**
+ * Calculates the credit needed to move some records between two nodes and
+ * adds this credit to @accum.
+ */
+static void btree_move_key_value_credit(struct m0_be_tx_credit *accum,
+					enum node_threshold th)
+{
+	/*depending on node_threshold, calculate fraction which needs to be
+	moved to other node*/
+	m0_bcount_t size = max_node_size*th/100;
```

Comment:
```
max_node_size * th / 100;
```
Space is needed between operand and operator.

Reply:
yes